python_modules/dagster/dagster/core/asset_defs/assets_job.py

```diff
@@ -207,42 +214,61 @@ def build_source_assets_by_key(
             source_assets_by_key[asset_source.key] = asset_source
         elif isinstance(asset_source, AssetsDefinition):
-            for asset_key, output_def in asset_source.output_defs_by_asset_key.items():
+            for output_name, asset_key in asset_source.asset_keys_by_output_name.items():
                 if asset_key:
-                    source_assets_by_key[asset_key] = output_def
+                    source_assets_by_key[asset_key] = asset_source.node_def.output_def_named(
+                        output_name
+                    )
 
     return source_assets_by_key
 
 
-def build_op_deps(
-    multi_asset_defs: Sequence[AssetsDefinition], source_paths: AbstractSet[AssetKey]
-) -> Dict[Union[str, NodeInvocation], Dict[str, IDependencyDefinition]]:
-    op_outputs_by_asset: Dict[AssetKey, Tuple[OpDefinition, str]] = {}
-    for multi_asset_def in multi_asset_defs:
-        for asset_key, output_def in multi_asset_def.output_defs_by_asset_key.items():
-            if asset_key in op_outputs_by_asset:
+def build_deps(
+    assets_defs: Sequence[AssetsDefinition], source_paths: AbstractSet[AssetKey]
+) -> Tuple[
+    Dict[Union[str, NodeInvocation], Dict[str, IDependencyDefinition]],
+    Mapping[NodeHandle, AssetsDefinition],
+]:
+    node_outputs_by_asset: Dict[AssetKey, Tuple[NodeDefinition, str]] = {}
+    assets_defs_by_node_handle: Dict[NodeHandle, AssetsDefinition] = {}
+
+    for assets_def in assets_defs:
+        for output_name, asset_key in assets_def.asset_keys_by_output_name.items():
+            if asset_key in node_outputs_by_asset:
                 raise DagsterInvalidDefinitionError(
                     f"The same asset key was included for two definitions: '{asset_key.to_string()}'"
                 )
 
-            op_outputs_by_asset[asset_key] = (multi_asset_def.op, output_def.name)
-
-    op_deps: Dict[Union[str, NodeInvocation], Dict[str, IDependencyDefinition]] = {}
-    for multi_asset_def in multi_asset_defs:
-        op_name = multi_asset_def.op.name
-        op_deps[op_name] = {}
-        for asset_key, input_def in multi_asset_def.input_defs_by_asset_key.items():
-            if asset_key in op_outputs_by_asset:
-                op_def, output_name = op_outputs_by_asset[asset_key]
-                op_deps[op_name][input_def.name] = DependencyDefinition(op_def.name, output_name)
-            elif asset_key not in source_paths and not input_def.dagster_type.is_nothing:
-                raise DagsterInvalidDefinitionError(
-                    f"Input asset '{asset_key.to_string()}' for asset '{op_name}' is not "
-                    "produced by any of the provided asset ops and is not one of the provided "
-                    "sources"
-                )
+            node_outputs_by_asset[asset_key] = (assets_def.node_def, output_name)
+
+    deps: Dict[Union[str, NodeInvocation], Dict[str, IDependencyDefinition]] = {}
+    collisions: Dict[str, int] = {}
```

Comment:
It would be helpful to provide some commenting and/or a more specific name for this dictionary to explain what's going on here: what are the keys of this dictionary? In what situations do we expect there to be collisions?